File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline

n = int(input())

for _ in range(n):
    w = list(input().rstrip())
    stk = []
    isVPS = True
    for i in w:
        if i == "(":
            stk.append(i)
        else:  # ")" 인 경우  - pop
            try:
                stk.pop()
            except:
                isVPS = False
                break
    if len(stk) == 0 and isVPS == True:
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.js
const filePath = process.platform === 'linux' ? 'dev/stdin' : './input.txt';
const input = require('fs').readFileSync(filePath).toString().split('\n');

const N = +input[0];

for (let i = 1; i <= N; i++) {
  const stack = [];
  let flag = true;

  for (const str of input[i]) {
    if (str === '(') {
      stack.push(str);
    } else {
      if (!stack.length) {
        flag = false;
        break;
      } else {
        stack.pop();
      }
    }
  }

  if (stack.length) flag = false;

  if (flag) console.log('YES');
  else console.log('NO');
}
